### PR TITLE
Do not show deleted users.

### LINKED ISSUE #191: Delete user
There is a button on the edit user page, which deletes the user.

 * The user deletion behavior is not defined.
 * There are no design for the user deletion page.

### PR DESCRIPTION
**Pull request by andorp**
Fixes #191.

### CODE REVIEW

**Comment by andorp** [comment]
I am going to squash the commits.

File: src/Frontend/Page/Admin.hs [hunk 1]
```diff
@@ -292,8 +292,19 @@ instance ToHtml PageAdminSettingsGaPUsersView where
...
-                let renderUserRow :: forall m. (Monad m) => User -> HtmlT m ()
-                    renderUserRow user = tr_ $ do
+                let renderUserRow :: forall m. (Monad m) => UserView -> HtmlT m ()
+                    renderUserRow (DeletedUser user) = tr_ $ do
+                        td_ $ nil
+                        td_ $ user ^. userLogin . unUserLogin . html
+                        td_ (case user ^. userRole of
+                                Student cl    -> toHtml $ showSchoolClass cl
+                                ClassGuest cl -> toHtml $ showSchoolClass cl
+                                _             -> nil)
```
Comment (fisx) [inline src/Frontend/Page/Admin.hs]: We have `ideaSpaceToSchoolClass` in `Types`.  This here could be a similar helper.  I thought we already had one, but I can't find it any more.

Comment (fisx) [inline src/Frontend/Page/Admin.hs]: (There is no reason to that in this PR, except that I just noticed now.)

File: src/Frontend/Page/Admin.hs [hunk 2]
```diff
@@ -292,16 +292,25 @@ instance ToHtml PageAdminSettingsGaPUsersView where
...
                         td_ (case user ^. userRole of
                                 Student cl    -> toHtml $ showSchoolClass cl
                                 ClassGuest cl -> toHtml $ showSchoolClass cl
                                 _             -> nil)
                         td_ $ roleLabel (user ^. userRole)
                         td_ (toHtmlRaw nbsp)
+
+                let renderUserRow :: forall m. (Monad m) => UserView -> HtmlT m ()
+                    renderUserRow (DeletedUser user) = tr_ $ do
+                        td_ nil
```
Comment (fisx) [inline src/Frontend/Page/Admin.hs]: Why not show the avatar of deleted users?  (Just curious, we can keep it this way.)

File: src/Types.hs [hunk 3]
```diff
@@ -105,6 +106,10 @@ instance (SOP.Generic a, SOP.Generic b, SOP.Generic c) => SOP.Generic (Either3 a
 app2 :: (c -> d) -> (a -> b -> c) -> a -> b -> d
 app2 f g x y = f $ g x y
 
+-- TODO: Check for application in the codebase
+fmap2 :: (Functor f, Functor g) => (a -> b) -> f (g a) -> f (g b)
+fmap2 f = fmap (fmap f)
```
Comment (fisx) [inline src/Types.hs]: We already have `Data.Functor.Infix.(<$$>)`.

File: src/Persistent/Idiom.hs [hunk 4]
```diff
@@ -127,3 +127,10 @@ setTopicPhase tid phase = modifyTopic tid $ topicPhase .~ phase
 deactivateUser :: AUID User -> AUpdate ()
 deactivateUser uid
     = modifyUser uid (userSettings . userSettingsPassword .~ UserPassDeactivated)
+
+getUserViews :: Query [UserView]
+getUserViews = fmap2 makeUserView getAllUsers
+
+findActiveUser :: AUID User -> EQuery User
+findActiveUser =
+    (maybe404 . (^? activeUser) . makeUserView) <=< maybe404 <=< findUser
```
Comment (fisx) [inline src/Persistent/Idiom.hs]: Why not define this in `MQuery`?  It seems to me like `maybe404` should only be called inside the application logic, when an `MQuery` is available but an `EQuery` is needed.

File: src/Action.hs [hunk 5]
```diff
@@ -219,7 +219,9 @@ loginByUser = login . view _Id
 loginByName :: (ActionPersist m, ActionUserHandler m) => UserLogin -> m ()
 loginByName n = do
     Just u <- query $ findUserByLogin n  -- FIXME: handle 'Nothing'
-    loginByUser u
+    if isDeletedUser u
+        then logout >> throwError500 "Unknown user identitifer"
+        else loginByUser u
```
Comment (fisx) [inline src/Action.hs]: We need to check this for `login`, `loginByUser` as well.  (I wonder if there is a better way to implement these three in terms of each other?)

File: src/Action.hs [hunk 6]
```diff
@@ -251,9 +253,11 @@ currentUser = do
     uid <- currentUserId
     muser <- query (findUser uid)
     case muser of
-        Just user -> pure user
-        Nothing   -> logout >> throwError500 "Unknown user identitifer"
+        Just user | isActiveUser user
+            -> pure user
+        _   -> logout >> throwError500 "Unknown user identitifer"
 
+-- TODO: Check if the current user is not deleted.
```
Comment (fisx) [inline src/Action.hs]: if the current user is deleted, something else has gone wrong!  (-:  So if you check it again here, it should be obvious that if the check fails it's an internal error.

File: src/Persistent/Idiom.hs (hunk 4 above)
Comment (andorp) [inline src/Persistent/Idiom.hs]: I would like to separate cases, but we don't have different exception for each cases, yet.

**Comment by fisx** [comment]
LGTM.  If you have processed all my comments and the remaining TODOs, you can merge.

File: src/Action.hs (hunk 6 above)
Comment (andorp) [inline src/Action.hs]: Race condition:

 * Somebody changes her profile
 * The admin deletes the user

I see cases when the user gets undeleted by the modification action.

File: src/Action.hs (hunk 5 above)
Comment (andorp) [inline src/Action.hs]: The `login` does not depend on the persistent, and `loginByUser` calls to login, let me tweak a little bit.

File: src/Frontend/Page/Admin.hs (hunk 2 above)
Comment (andorp) [inline src/Frontend/Page/Admin.hs]: The difference on the UI was more pleasant to the eye.

File: src/Frontend/Page/Admin.hs (hunk 1 above)
Comment (np) [inline src/Frontend/Page/Admin.hs]: You can use the roleSchoolClass lens: `td_ $ user ^. userRole . roleSchoolClass . to showSchoolClass . html`

File: src/Action.hs (hunk 6 above)
Comment (fisx) [inline src/Action.hs]: > I see cases when the user gets undeleted by the modification action.

good point!

**Comment by np** [comment]
LGTM

**Comment by andorp** [comment]
Squashed.

**Comment by fisx** [comment]
Thanks!

**Pull request merged**